raiden/blockchain/events.py

```diff
@@ -246,18 +249,40 @@ def event_to_state_change(event):  # pylint: disable=too-many-return-statements
 class BlockchainEvents(object):
     """ Events polling. """
 
-    def __init__(self):
+    def __init__(self, chain, address, registry_address):
         self.event_listeners = list()
+        self.chain = chain
+        self.address = address
+        self.registry_address = registry_address
 
     def poll_all_event_listeners(self):
         result = list()
+        reinstalled_filters = False
 
-        for event_listener in self.event_listeners:
-            decoded_events = poll_event_listener(
-                event_listener.filter,
-                event_listener.translator,
-            )
-            result.extend(decoded_events)
+        while True:
+            try:
+                for event_listener in self.event_listeners:
+                    decoded_events = poll_event_listener(
+                        event_listener.filter,
+                        event_listener.translator,
+                    )
+                    result.extend(decoded_events)
+                break
+            except EthNodeCommunicationError as e:
+                # If the eth client has restarted and we reconnected to it then
+                # filters will no longer exist there. In that case we will need
+                # to recreate all the filters.
+                if not reinstalled_filters and str(e) == 'filter not found':
+                    self.event_listeners = list()
+                    result = list()
+                    reinstalled_filters = True
+                    self.add_proxies_listeners(get_relevant_proxies(
```

Comment:
Not sure, if recreating the proxy instances here is the right approach (i.e. calling `get_relevant_proxies`). I suspect that we double the proxy instances by this approach (i.e. the old instances may still be referenced from some other node).

Reply:
I think konrad is correct here. Perhaps try to see where we save the data you need when we install proxies at [raiden_service](https://github.com/raiden-network/raiden/blob/master/raiden/raiden_service.py#L714) and extract it from there?

Reply:
Can you try to re-install from `self.event_listeners`, or, if they don't hold enough information, extend them to allow for re-installation? That way we also won't have the potentially cyclic dependency of `self.chain` in this class.

Reply:
I think it's fine to reinstall the filters. Since filters are *not* persistent, once the node is restarted and the id from the previous run is used the error will be `filter not found`, reinstalling the filters seems fine under this circumstance.

@LefterisJP The filters are kept inside the `BlockchainEvents` only.

Reply:
There are problems with this patch though.

1. It may miss events under some race conditions. This will happen if the filters miss a block before being re-installed.
2. It's passing the `chain` to yet another object, this may be better handled by something else.

Example for 1.:

```
- Block 5
- Events are polled
- The node goes offline
- Block 6
- Block 7
- The filters are reinstalled
```

Under the above scenario, events from the block 6 will be lost. Setting [`fromBlock{ ]](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_newfilter) with the latest polled block should be fine.

Note: Processing the same event multiple times is *not* a problem, events must be idempotent.

Reply:
Any recommendations on how to handle point 2? I'm not sure how to architect this in a nice way.